### PR TITLE
Feature/calendarevents

### PR DESCRIPTION
**Pull request by TamaroWalter**
Pull request for issue #2 :
Adds the main functionality to the townsquare block
Adds tests for the first stable version

### CODE REVIEW

File: db/access.php [hunk 1]
```diff
@@ -17,28 +17,30 @@
...
         'contextlevel' => CONTEXT_SYSTEM,
         'archetypes' => [
-            'user' => CAP_ALLOW
+            'user' => CAP_ALLOW,
         ],
-        'clonepermissionsfrom' => 'moodle/my:manageblocks'
+        'clonepermissionsfrom' => 'moodle/my:manageblocks',
     ],
+
     'block/towwnsquare:addinstance' => [
```
[the remark this answers is not in the document]
Comment (TamaroWalter) [reply db/access.php:36]: Well,  `myaddinstance` is to add it to the dashboard, every user should have the capability.
`addinstance` is to add the block to a course page. As the plugin should only work in the dashboard, even editingteachers and managers should not be able to add it to a course

File: templates/postletter.mustache [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+                    {{/str}}
+                {{/author}}
+            </bold>
+        </div>
+        <p></p>
+        <div class = "postletter_content">
+                <span class ="postletter_message" id="content-nr-{{contentid}}" >
+                    {{{message}}}
+                </span>
+                <span class="townsquare_showmore badge rounded-pill bg-light text-dark" id="content-nr-{{contentid}}"
```
[the remark this answers is not in the document]
Comment (TamaroWalter) [reply templates/postletter.mustache:60]: Yes, the 'text-dark' prevents that the text can be shown as a light color in the dark mode. It's deleted now

File: classes/townsquareevents.php [hunk 3]
```diff
@@ -0,0 +1,339 @@
...
+
+        // Extend the strings for the 2 module cases.
+        if ($modulename == 'forum') {
+            $begin .= "'forum' AS modulename, module.id AS forumid,";
+            $middle .= "FROM {forum_posts} posts
+                        JOIN {forum_discussions} discuss ON discuss.id = posts.discussion
+                        JOIN {forum} module ON module.id = discuss.forum ";
+        } else if ($modulename == 'moodleoverflow') {
+            $begin .= "'moodleoverflow' AS modulename, module.id AS moodleoverflowid, ";
+            $middle .= "FROM {moodleoverflow_posts} posts
```
[the remark this answers is not in the document]
Comment (TamaroWalter) [reply classes/townsquareevents.php:249]: Is checked and testet :)

File: classes/townsquareevents.php [hunk 4]
```diff
@@ -411,4 +426,20 @@ private function add_anonymousattribute($posts): array {
...
+    /**
+     * Filter that checks if the event is available for the current user.
+     * Applies to restriction that are defined in the module setting (restrict access).
+     * @param object $event The event that is checked.
+     * @return bool if the event is available for the current user.
+     */
+    private function filter_availability($event): bool {
+        // If there is a restriction, check if it applies to the user.
+        $modinfo = get_fast_modinfo($event->courseid);
+        $cm = $modinfo->get_cm($event->coursemoduleid);
```
Comment (NinaHerrmann) [inline classes/townsquareevents.php:445]: Do you know what additional information is stored in $cm?

File: .github/workflows/moodle-ci.yml [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+
+      - name: Initialise moodle-plugin-ci
+        run: |
+          composer create-project -n --no-dev --prefer-dist moodlehq/moodle-plugin-ci ci ^3
+          echo $(cd ci/bin; pwd) >> $GITHUB_PATH
+          echo $(cd ci/vendor/bin; pwd) >> $GITHUB_PATH
+          sudo locale-gen en_AU.UTF-8
+          echo "NVM_DIR=$HOME/.nvm" >> $GITHUB_ENV
+
+      - name: Install tool plugin
```
Comment (NinaHerrmann) [inline .github/workflows/moodle-ci.yml:59]: mod

File: block_townsquare.php [hunk 6]
```diff
@@ -14,45 +14,46 @@
...
         }
 
-        $this->content = new stdClass();
-        $this->content->footer = '';
-
-        // Add logic here to define your template data or any other content.
-        $data = ['YOUR DATA GOES HERE'];
-
-        $this->content->text = $OUTPUT->render_from_template('block_townsquare/content', $data);
+        $controller = new contentcontroller();
```
[the remark this answers is not in the document]
Comment (NinaHerrmann) [reply block_townsquare.php:49]: This was not resolved please check it again or reason why you are not taking the feedback!

**Pull request merged**